app/models/application_record.rb

```diff
@@ -1,3 +1,29 @@
 class ApplicationRecord < ActiveRecord::Base
   self.abstract_class = true
+
+  after_create { EventContext.trigger_event(:entity_created, event_data) }
+  after_update { EventContext.trigger_event(:entity_updated, event_data) unless saved_change_to_attribute?(:last_activity_at) }
+  after_destroy { EventContext.trigger_event(:entity_destroyed, event_data) }
+
+  private
+
+  def event_data
+    { entity: self.class.name }.merge(anonymised_attributes)
+  end
+
+  def anonymised_attributes
+    attributes.each_with_object({}) do |(key, value), anonymised|
+      next unless export?(key)
+
+      anonymised[key] = anonymise?(key) ? StringAnonymiser.new(value).to_s : value
```

Comment:
I know in some places anonymised data was suffixed with `-identifier`. If @stevenleggdfe is happy we should change those to be consistent and just send the attribute key without `-identifier` everywhere